src/modules/zfs/ZfsJob.cpp

```diff
@@ -0,0 +1,326 @@
+        poolNameEntry[ "poolName" ] = poolName;
+        poolNameEntry[ "mountpoint" ] = mountpoint;
+        poolNameEntry[ "dsName" ] = "none";
+
+        // If the mountpoint is /, create datasets per the config file. If not, create a single dataset mounted at the partitions mountpoint
+        if ( mountpoint == '/' )
+        {
+            CollectMountpoints( partitions );
+            QVariantList datasetList;
+            for ( const auto& dataset : qAsConst( m_datasets ) )
+            {
+                QVariantMap datasetMap = dataset.toMap();
+
+                // Make sure all values are valid
+                if ( datasetMap[ "dsName" ].toString().isEmpty() || datasetMap[ "mountpoint" ].toString().isEmpty()
+                     || datasetMap[ "canMount" ].toString().isEmpty() )
+                {
+                    cWarning() << "Bad dataset entry";
+                    continue;
+                }
+
+                // We should skip this dataset if it conflicts with a permanent mountpoint
+                if ( IsMountpointOverlapping( datasetMap[ "mountpoint" ].toString() ) )
+                {
+                    continue;
+                }
+
+                // Create the dataset.  We set canmount=no regardless of the setting for now.
+                // It is modified to the correct value in the mount module to ensure mount order is maintained
+                auto r = system->runCommand( { "sh",
+                                               "-c",
+                                               "zfs create " + m_datasetOptions + " -o canmount=off -o mountpoint="
+                                                   + datasetMap[ "mountpoint" ].toString() + " " + poolName + "/"
+                                                   + datasetMap[ "dsName" ].toString() },
+                                             std::chrono::seconds( 10 ) );
+                if ( r.getExitCode() != 0 )
+                {
+                    cWarning() << "Failed to create dataset" << datasetMap[ "dsName" ].toString();
+                    continue;
+                }
+
+                // Add the dataset to the list for global storage this information is used later to properly set
+                // the mount options on each dataset
+                datasetMap[ "zpool" ] = m_poolName;
+                datasetList.append( datasetMap );
+            }
+
+            // If the list isn't empty, add it to global storage
+            if ( !datasetList.isEmpty() )
+            {
+                gs->insert( "zfsDatasets", datasetList );
+            }
+        }
+        else
+        {
+            // This is a zpool with a single dataset We again set canmount=no regardless of the desired setting.
+            // It is modified to the correct value in the mount module to ensure mount order is maintained
+            QString dsName = mountpoint;
+            dsName = AlphaNumeric( mountpoint );
+            auto r = system->runCommand( { "sh",
```

Comment:
Sort of a general question, why all the `sh -c` calls? To me it looks like this could be done with less string-manipulations and more building-a-command-line-as-a-QStringList.

eg. here,
```
QStringList cmd{ "zfs", "create" };
cmd << m_datasetOptions; // Probably needs actual work, since this might be a bunch of space-separated `-o opt=val` strings
cmd << "-o" << "canmount=off" << "-o" << QStringLiteral("mountpoint=%1").arg(mountpoint);
cmd << (poolName + "/" + dsName);
```

Hm, now that I write it like that I'm not **really** keen on how readable it becomes with streaming operators :(

Reply:
.. looking at the config file, much later, there's string-valued configuration options that get substituted in, which might be anything; that makes the shell-because-command-line-parsing necessary (or we need to invest in `QStringList::split(' ')`)

Reply:
> Sort of a general question, why all the `sh -c` calls? 

I tried to only use `sh -c` when it was needed.  i.e. When I was using shell pipelines.

However, in this case and the one above at line 254, it doesn't look needed to me. I will convert it and try to see if I can make it readable.